system/jlib/jsort.cpp

```diff
@@ -703,6 +781,212 @@ void msortvecstableinplace(void ** rows, size32_t n, const ICompare & compare, v
+            return NULL;
+        }
+
+    protected:
+        const ICompare & compare;
+        void * * result;
+        void * * src1;
+        void * * src2;
+        size_t n1;
+        size_t n2;
+        size_t n;
+    };
+
+    class MergeRevTask : public MergeTask
+    {
+    public:
+        MergeRevTask(const ICompare & _compare, void * * _result, size_t _n1, void * * _src1, size_t _n2, void * * _src2, size_t _n)
+        : MergeTask(_compare, _result, _n1, _src1, _n2, _src2, _n)
+        {
+        }
+
+        virtual task * execute()
+        {
+            mergePartitionsRev(compare, result, n2, src2, n1, src1, n);
+            return NULL;
+        }
+    };
+
+public:
+    TbbParallelMergeSorter(void * * _rows, const ICompare & _compare) : compare(_compare), baseRows(_rows)
+    {
+        //The following constants control the number of iterations to be performed in parallel.
+        //The sort is split into more parts than there are cpus so that the effect of delays one task tend to be evened out.
+        //The following constants should possibly be tuned on each platform.  The following gave a good balance on a 2x8way xeon
+        const unsigned extraBisectDepth = 3;
+        const unsigned extraParallelMergeDepth = 3;
+
+        unsigned numCpus = tbb::task_scheduler_init::default_num_threads();
+        unsigned ln2NumCpus = (numCpus <= 1) ? 0 : getMostSignificantBit(numCpus-1);
+        assertex(numCpus <= (1U << ln2NumCpus));
+
+        //Merge in parallel once it is likely to be beneficial
+        parallelMergeDepth = ln2NumCpus+ extraParallelMergeDepth;
+        //Aim to execute in parallel until the width is 8*the maximum number of parallel task
+        singleThreadDepth = ln2NumCpus + extraBisectDepth;
+    }
+    ~TbbParallelMergeSorter()
+    {
+    }
+
+    void sortRoot(void ** rows, size32_t n, void ** temp)
+    {
+        task * end = new (task::allocate_root()) tbb::empty_task();
+        end->set_ref_count(1+1);
+        task * task = new (task::allocate_root()) BisectTask(*this, rows, n, temp, 0, end);
+        end->spawn(*task);
+        end->wait_for_all();
+        end->destroy(*end);
+    }
+
+public:
+    const ICompare & compare;
+    unsigned singleThreadDepth;
+    unsigned parallelMergeDepth;
+    void * * baseRows;
+};
+
+//-------------------------------------------------------------------------------------------------------------------
+void parmsortvecstableinplace(void ** rows, size32_t n, const ICompare & compare, void ** temp, unsigned ncpus)
+{
+    if ((n <= singleThreadedMSortThreshold) || ncpus == 1)
+    {
+        msortvecstableinplace(rows, n, compare, temp);
+        return;
+    }
+
+    TbbParallelMergeSorter sorter(rows, compare);
+    sorter.sortRoot(rows, n, temp);
+}
+#else
+void parmsortvecstableinplace(void ** rows, size32_t n, const ICompare & compare, void ** temp, unsigned ncpus)
+{
+    parqsortvecstableinplace(rows, n, compare, temp, ncpus);
```

Comment:
It could be.  I could make that change in another request (adding varients to roxie jira).

We tend to fallback to a different implementation in the engines if the version they request isn't available.